proto/iotics/api/meta.proto

```diff
@@ -0,0 +1,125 @@
+  SPARQL_CSV = 2;
+  // Applicable to CONSTRUCT/DESCRIBE (Terse RDF Triple Language)
+  RDF_TURTLE = 3;
+  // Applicable to CONSTRUCT/DESCRIBE (RDF 1.1 XML)
+  RDF_XML = 4;
+  // Applicable to CONSTRUCT/DESCRIBE (RDF 1.1 N-Triples)
+  RDF_NTRIPLES = 5;
+  // Applicable to CONSTRUCT/DESCRIBE (RDF 1.1 N-Quads)
+  RDF_NQUADS = 6;
+}
+
+// SparqlQueryRequest describes a SPARQL query.
+message SparqlQueryRequest {
+
+  // SPARQL request payload.
+  message Payload {
+
+    // The desired result content type. Note that the actual content (mime) type could be different and will be returned
+    // in result chunks.
+    SparqlResultType resultTypeHint = 1;
+    // A UTF8-encoded SPARQL 1.1 query
+    bytes query = 2;
+  }
+
+  // SPARQL request headers
+  Headers headers = 1;
+
+  // SPARQL request scope
+  Scope scope = 2;
+
+  // SPARQL request payload
+  Payload payload = 3;
+}
+
+// Part of a result for a SPARQL query request. Multiple chunks form a complete result. Related chunks can be identified
+// by a combination of:
+// - The host ID (unset for local results)
+// - Client reference (in headers, set by caller)
+// - Chunk sequence number
+// Note that chunks are guaranteed to arrive in-order (for a specific result from a host) but could be duplicated. It is
+// up to the client to ignore duplicates (based on sequence number).
+message SparqlQueryResultChunk {
+
+  // Payload of the query result chunk
+  message Payload {
+
+    // Result host identifier. Indicates from which host this result chunk came from. For a local result, this field
+    // will be unset.
+    HostID remoteHostId = 1;
+
+    // Position of a chunk in result from a given host (and request). The first chunk has a sequence number of 0.
+    uint64 seqNum = 2;
+
+    // Indicates whether this is the last chunk from a given host, for a specific request. Results for different
+    // requests can be identified by setting a unique clientRef in the request headers.
+    bool last = 3;
+
+    // Result error status (only applicable to local results, i.e. when remoteHostId is unset). If set, this will
+    // indicate a problem with running the query (e.g. invalid syntax) as opposed to a more general issue (in which case
+    // the standard gRPC error mechanism will be used and the stream terminated).
+    google.rpc.Status status = 4;
+
+    // The actual result content type (as raw mime type). For the hinted type (in a request), typically the returned
+    // type can be:
+    // - SPARQL_JSON - application/sparql-results+json
+    // - SPARQL_XML - application/sparql-results+xml
+    // - SPARQL_CSV - text/csv
+    // - RDF_TURTLE - text/turtle
+    // - RDF_XML - application/rdf+xml
+    // - RDF_NTRIPLES - application/n-triples
+    // - RDF_NQUADS - application/n-quads
+    string actualType = 5;
+
+    // Query result chunk, encoded according to actualType. The maximum size of each chunk is host-specific.
```

Comment:
if changing actualType, update comment here.

Reply:
"The maximum size of each chunk is host-specific." - Does this mean a rogue host could DOS the network with massive chunk settings? Is there an absolute maximum?

Reply:
Should this be a value specified in the request, up until the max value set by host?

Reply:
> Should this be a value (chunkSize) specified in the request, up until the max value set by host?

We could offer that - but then we'd also definitely want a minimum value (sending tiny chunks is also a type of DoS - one has to send many more messages/packets/whatever than with larger ones).
So maybe it should be a chunk size hint. Note that a remote host might choose to respond with a differently sized chunk. This would mean that locally one might have to split (or combine) chunks based on the client's suggestion.
Maybe a "max chunk size" client option might be best: "_I can handle anything up to at most this size_"

> Does this mean a rogue host could DOS the network with massive chunk settings? Is there an absolute maximum?

Chunks apply to responses, so that would mean a remote host would send massive chunks to the originating (local) host (i.e. a point-to-point connection). We have the option (not implemented yet) to reject oversized payloads from the p2p network - that would be a general issue - not specific to this type of request. It's a good one to look into but I think that is a general "hardening the host" issue.
@smartrics  - do you already have something in the road-map about making hosts robust when communicating with other peers/hosts?

Reply:
Yes, I agree it's definitely a more general concern than just this request.

Reply:
No on the roadmap yet but we will have to consider Network API security as soon as we will offer on premise deployment
- host twin authenticity 
- host banishment
- signed payload
- ....

Reply:
> This would mean that locally one might have to split (or combine) chunks based on the client's suggestion.

The client might already expect different chunk sizes since it is configurable by host? :eyes:

Reply:
> The client might already expect different chunk sizes since it is configurable by host?

@StephanieBracaloni I don't understand what you're saying. Do you mean: Chunk size should be configurable in host, **not** per-request? (The original proposal.)

Note that grpc ["max receive message size"](https://github.com/grpc/grpc-go/blob/master/rpc_util.go#L572-L573) is a client-only option. I'm not sure we should let the client set an expectation. If it receives something too large it doesn't want to handle, it can ignore it. (Otherwise we're adding more logic / negotiation )

Maximum size is something that also applies to:
- search results (if had a higher limit than now)
- twin/feed listing (again, if configured higher limit)
- feed share receive

So it feels like there should be a general solution (at least in that we consider it as such) first before going down the route of request-specific flags.

Reply:
I was thinking about a chunk size per host for the first approach.
Then we can also offer the option to the end-user if needed (with a max value in the host).

In both cases, the end-user should expect to receive different chunk sizes since the remote request will come from a remote host with a different configuration.